### PR TITLE
Feature: expanding migration of the cover with the submitting a claim

### PR DESCRIPTION
**Pull request by MilGard91**
## Context

Currently submitting a claim for V1 cover takes 2 steps, for easier integration with FE this can all be done in one tx.

Issue: #496 

## Changes proposed in this pull request

Expanding the migration method with `submitClaim` function from IndividualClaims contract. For this to work, CoverMigratot contract need to be added as member so it can hold and transfer CoverNFTs.

## Test plan

Tests expanded with integration test for this process, also new Mock Contract added for easier adding V1 Claims.

Removing old unit test, since it just stored data.

## Checklist

- [x] Rebased the base branch
- [x] Attached corresponding Github issue
- [x] Prefixed the name with the type of change (i.e. feat, chore, test)
- [x] Performed a self-review of my own code
- [x] Followed the style guidelines of this project
- [ ] Made corresponding changes to the documentation
- [x] Didn't generate new warnings
- [x] Didn't generate failures on existing tests
- [x] Added tests that prove my fix is effective or that my feature works

## Review

When reviewing a PR, please indicate intention in comments using the following emojis:
* :cake: = Nice to have but not essential.
* :bulb: = Suggestion or a comment based on personal opinion
* :hammer: = I believe this should be changed.
* :thinking: = I don’t understand something, do you mind giving me more context?
* :rocket: = Feedback

### CODE REVIEW

File: contracts/modules/cover/CoverMigrator.sol [hunk 1]
```diff
@@ -20,16 +20,33 @@ contract CoverMigrator is MasterAwareV2 {
...
-  function submitClaim(uint coverId) external whenNotPaused {
-    cover().migrateCoverFromOwner(coverId, msg.sender, tx.origin);
+  function submitClaim(
+    uint32 coverId,
+    uint16 segmentId,
+    uint96 requestedAmount,
+    string calldata ipfsMetadata
+  ) payable external whenNotPaused returns (uint newCoverId){
+    newCoverId =  cover().migrateCoverFromOwner(coverId, msg.sender, address(this));
+    individualClaims().submitClaim{value: msg.value}(uint32(newCoverId), segmentId, requestedAmount, ipfsMetadata);
```
Comment (danoctavian) [inline contracts/modules/cover/CoverMigrator.sol:41]: Can This cover ever be yield token cover in which case this doesn't work?

File: contracts/modules/cover/CoverMigrator.sol [hunk 2]
```diff
@@ -1,6 +1,6 @@
 // SPDX-License-Identifier: GPL-3.0-only
 
-pragma solidity ^0.8.9;
+pragma solidity ^0.8.16;
```
Comment (danoctavian) [inline contracts/modules/cover/CoverMigrator.sol:3]: I would like a comment on this file that reminds us this contract will replace 

https://github.com/NexusMutual/smart-contracts/blob/master/contracts/modules/claims/Claims.sol#L82


And the signature of submitClaims needs to match exactly for compatibility.

File: contracts/modules/cover/CoverMigrator.sol [hunk 3]
```diff
@@ -20,16 +20,33 @@ contract CoverMigrator is MasterAwareV2 {
...
+  function coverNFT() internal returns (ICoverNFT) {
+    return ICoverNFT(cover().coverNFT());
+  }
+
   /// @dev Migrates covers for arNFT-like contracts that don't use Gateway.sol
   ///
   /// @param coverId          Legacy (V1) cover identifier
-  function submitClaim(uint coverId) external whenNotPaused {
-    cover().migrateCoverFromOwner(coverId, msg.sender, tx.origin);
+  function submitClaim(
```
Comment (danoctavian) [inline contracts/modules/cover/CoverMigrator.sol:34]: I don't expect this to work anymore given it's original purpose.

https://github.com/NexusMutual/smart-contracts/blob/master/contracts/modules/claims/Claims.sol#L82

CoverMigrator was replacing Claims and whatever contracts would call submitClaim on Claims.sol would now hit this function.

The signature is now changed so it won't work.

We can have 2 separate functions here if this new one still helps with UX in some cases.

File: contracts/modules/cover/CoverUtilsLib.sol [hunk 4]
```diff
@@ -46,17 +46,23 @@ library CoverUtilsLib {
     ProductType[] storage _productTypes,
     CoverData[] storage _coverData,
     mapping(uint => CoverSegment[]) storage _coverSegments
-  ) external {
+  ) external returns(uint newCoverId){
```
Comment (danoctavian) [inline contracts/modules/cover/CoverUtilsLib.sol:49]: formatting is wrong here there's spaces before and after those parentheses

File: test/integration/Cover/migrateCoverAndSubmit.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+  return new Uint8Array(length).fill(0).map((x, i) => bytes[i] || x);
+}
+
+describe('createStakingPool', function () {
+  let coverOwner;
+  let coverId;
+  before(async function () {
+    const ETH = zeroPadRight(Buffer.from('ETH'), 4);
+    const { qd, tk: token } = this.contracts;
+    coverOwner = this.accounts.members[1];
```
Comment (danoctavian) [inline test/integration/Cover/migrateCoverAndSubmit.js:17]: create some spaces here between contract initialization, account initialization and the rest of the code.

Use more spaces when you write test code to make things easier to read and group things logically.

For example see how these tests are written:

https://github.com/NexusMutual/smart-contracts/blob/d737f8f28d9797c1a2a415d51a8337b1b1e5bdf1/test/unit/IndividualClaims/redeemClaimPayout.js#L56

File: test/integration/Cover/migrateCoverAndSubmit.js [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+const { expect } = require('chai');
+const { getCoverSegment } = require('../../unit/IndividualClaims/helpers');
+
+const { parseEther } = ethers.utils;
+
+function zeroPadRight(bytes, length) {
+  return new Uint8Array(length).fill(0).map((x, i) => bytes[i] || x);
+}
+
+describe('createStakingPool', function () {
```
Comment (danoctavian) [inline test/integration/Cover/migrateCoverAndSubmit.js:11]: this is called createStakingPool

File: test/unit/CoverMigrator/setup.js [hunk 7]
```diff
@@ -1,50 +0,0 @@
-const { ethers } = require('hardhat');
```
Comment (danoctavian) [inline test/unit/CoverMigrator/setup.js:1]: what's the reason for removing the CoverMigrator tests?

File: test/integration/utils/enroll.js [hunk 8]
```diff
@@ -23,6 +23,16 @@ async function enrollMember({ mr, tk, tc }, members, kycAuthSigner, options = {}
     await tk.connect(member).approve(tc.address, ethers.constants.MaxUint256);
     await tk.transfer(member.address, initialTokens);
   }
+  const membershipApprovalData0 = await signMembershipApproval({
```
Comment (danoctavian) [inline test/integration/utils/enroll.js:26]: The top level function is called enrollMember and enrolls a set of members (could be better called enrollMembers but nvm for now)

Why does it run an extra routine here to enroll one of the contracts a member?

This logic should not exist here from the point of view of what this function should do

This logic should exist wherever this registration of cl is necessary but not in this commonly used function for enrolling members.

File: contracts/modules/assessment/IndividualClaims.sol [hunk 9]
```diff
@@ -315,7 +315,7 @@ contract IndividualClaims is IIndividualClaims, MasterAwareV2 {
       (
         bool refunded,
         /* bytes data */
-      ) = msg.sender.call{value: msg.value - assessmentDepositInETH}("");
+      ) = tx.origin.call{value: msg.value - assessmentDepositInETH}("");
```
Comment (danoctavian) [inline contracts/modules/assessment/IndividualClaims.sol:318]: what's the context for this change?

Completely changes semantics of returning the deposit

**Review by danoctavian — CHANGES_REQUESTED** [review]
about CoverMigrator submitClaim and it's use.
 
 Unless our specs have changed this current approach would break compatibility with existing contracts doing submitClaim.
 
 Other issues in the testing code.
 
 The other issue standing out is why the deposit return in Assessment is now done to tx.origin

File: contracts/modules/assessment/IndividualClaims.sol (hunk 9 above)
Comment (danoctavian) [reply contracts/modules/assessment/IndividualClaims.sol:318]: As discussed offband this change in not necessary or appropriate.

We should use msg.sender and proxy the ET back and forth to the original msg.sender.

For all intents and purposes IndividualClaims.submitClaim should only refer to msg.sender.

File: contracts/modules/cover/Cover.sol [hunk 10]
```diff
@@ -150,8 +151,20 @@ contract Cover is ICover, MasterAwareV2, IStakingPoolBeacon, ReentrancyGuard {
...
+  ) external override onlyInternal returns (uint newCoverId) {
+    newCoverId = _migrateCoverFromOwner(coverId, fromOwner, newOwner);
+  }
+
+  function migrateAndSubmitClaim(
+    uint32 coverId,
+    uint16 segmentId,
+    uint96 requestedAmount,
+    string calldata ipfsMetadata
+  ) payable external whenNotPaused returns (uint newCoverId){
```
Comment (danoctavian) [inline contracts/modules/cover/Cover.sol:163]: formatting issue here: ){ no space

File: contracts/mocks/Legacy/MockLegacyQuotationData.sol [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+  external
+  {
+    uint expiryDate = now.add(uint(_coverPeriod).mul(1 days));
+    allCovers.push(Cover(_userAddress, _currencyCode,
+      _sumAssured, _coverPeriod, expiryDate, _scAddress, premiumNXM));
+    uint cid = allCovers.length.sub(1);
+    userCover[_userAddress].push(cid);
+    emit CoverDetailsEvent(cid, _scAddress, _sumAssured, expiryDate, premium, premiumNXM, _currencyCode);
+  }
+}
```
Comment (danoctavian) [inline contracts/mocks/Legacy/MockLegacyQuotationData.sol:35]: make sure there's no new line here

File: contracts/modules/assessment/IndividualClaims.sol [hunk 12]
```diff
@@ -230,12 +230,31 @@ contract IndividualClaims is IIndividualClaims, MasterAwareV2 {
...
+    _submitClaim(coverId, segmentId, requestedAmount, ipfsMetadata, owner);
+  }
+
+  function _submitClaim(
+    uint32 coverId,
+    uint16 segmentId,
+    uint96 requestedAmount,
+    string calldata ipfsMetadata,
+    address owner
+  ) internal whenNotPaused returns (Claim memory) {
```
Comment (danoctavian) [inline contracts/modules/assessment/IndividualClaims.sol:253]: no need for whenNotPaused here, it's an internal function, the care is for the publicly exposed functions

File: contracts/modules/assessment/IndividualClaims.sol [hunk 13]
```diff
@@ -230,12 +230,31 @@ contract IndividualClaims is IIndividualClaims, MasterAwareV2 {
...
+  require(
       coverNFT.isApprovedOrOwner(msg.sender, coverId),
-      "Only the owner or approved addresses can submit a claim"
+      "The owner or approved address can submit a claim"
```
Comment (danoctavian) [inline contracts/modules/assessment/IndividualClaims.sol:256]: i think the old text of the revert was fine. it can be either an owner or an approved address, so there's multiple addresses

File: contracts/modules/assessment/IndividualClaims.sol [hunk 14]
```diff
@@ -230,12 +230,31 @@ contract IndividualClaims is IIndividualClaims, MasterAwareV2 {
     uint16 segmentId,
     uint96 requestedAmount,
     string calldata ipfsMetadata
-  ) external payable override onlyMember whenNotPaused returns (Claim memory) {
-    require(
+  ) external payable override onlyMember whenNotPaused returns (Claim memory claim) {
+    claim = _submitClaim(coverId, segmentId, requestedAmount, ipfsMetadata, msg.sender);
```
Comment (danoctavian) [inline contracts/modules/assessment/IndividualClaims.sol:234]: just return _submitClaim, it's redundant to have the variable named

File: contracts/modules/assessment/IndividualClaims.sol (hunk 13 above)
Comment (MilGard91) [reply contracts/modules/assessment/IndividualClaims.sol:256]: I got a warning that the contract exceeded the limit, so I had to change few  stuff so it would pass the check...

File: contracts/modules/assessment/IndividualClaims.sol [hunk 15]
```diff
@@ -230,12 +230,31 @@ contract IndividualClaims is IIndividualClaims, MasterAwareV2 {
...
+    return claim;
+  }
+  function submitClaimOf(
+    uint32 coverId,
+    uint16 segmentId,
+    uint96 requestedAmount,
+    string calldata ipfsMetadata,
+    address owner
+  ) external payable override onlyInternal whenNotPaused {
+    _submitClaim(coverId, segmentId, requestedAmount, ipfsMetadata, owner);
```
Comment (danoctavian) [inline contracts/modules/assessment/IndividualClaims.sol:244]: let's return the claim here as well to make things more flexible for future use

File: contracts/modules/cover/CoverMigrator.sol [hunk 16]
```diff
@@ -21,6 +21,8 @@ contract CoverMigrator is MasterAwareV2 {
   }
 
   /// @dev Migrates covers for arNFT-like contracts that don't use Gateway.sol
+  ///      The function should not be changed in anyway, because it is being used instead of the old Claims.sol method
```
Comment (danoctavian) [inline contracts/modules/cover/CoverMigrator.sol:24]: should not be changed in any way* (two separate words)

File: contracts/modules/assessment/IndividualClaims.sol (hunk 13 above)
Comment (danoctavian) [reply contracts/modules/assessment/IndividualClaims.sol:256]: are you running with `export ENABLE_OPTIMIZER=1 ; npm run compile` ?
 
 not expecting this to occur with IC.

File: contracts/interfaces/IMasterAwareV2.sol [hunk 17]
```diff
@@ -13,7 +13,8 @@ interface IMasterAwareV2 {
     AS, // Assessment.sol
     TK, // NXMToken.sol
     PS, // LegacyPooledStaking.sol
-    GV  // Governance.sol
+    GV,  // Governance.sol
+    IC  // IndividualClaims.sol
```
Comment (shark0der) [inline contracts/interfaces/IMasterAwareV2.sol:17]: I think we may have a conflict here. `IC` is used in production for the `Incidents`. IIRC Dragos intended to "take over" the abbreviation and reuse it for the other contract but I'm really unsure here.

File: test/unit/index.js [hunk 18]
```diff
@@ -7,7 +7,6 @@ describe('UNIT TESTS', function () {
   require('./MCR');
   require('./Assessment');
   require('./Cover');
-  require('./CoverMigrator');
```
Comment (shark0der) [inline test/unit/index.js:10]: Why was this removed?

File: contracts/modules/assessment/IndividualClaims.sol [hunk 19]
```diff
@@ -230,12 +230,30 @@ contract IndividualClaims is IIndividualClaims, MasterAwareV2 {
     uint16 segmentId,
     uint96 requestedAmount,
     string calldata ipfsMetadata
-  ) external payable override onlyMember whenNotPaused returns (Claim memory) {
-    require(
+  ) external payable override onlyMember whenNotPaused returns (Claim memory claim) {
+    return _submitClaim(coverId, segmentId, requestedAmount, ipfsMetadata, msg.sender);
+  }
+  function submitClaimOf(
```
Comment (shark0der) [inline contracts/modules/assessment/IndividualClaims.sol:236]: Make sure there's a new line between the functions

File: contracts/modules/assessment/IndividualClaims.sol [hunk 20]
```diff
@@ -230,12 +230,30 @@ contract IndividualClaims is IIndividualClaims, MasterAwareV2 {
...
+  }
+
+  function _submitClaim(
+    uint32 coverId,
+    uint16 segmentId,
+    uint96 requestedAmount,
+    string calldata ipfsMetadata,
+    address owner
+  ) internal returns (Claim memory) {
+  require(
```
Comment (shark0der) [inline contracts/modules/assessment/IndividualClaims.sol:253]: Missed the indent here.

File: contracts/modules/assessment/IndividualClaims.sol [hunk 21]
```diff
@@ -308,14 +326,13 @@ contract IndividualClaims is IIndividualClaims, MasterAwareV2 {
...
 
-
     require(msg.value >= assessmentDepositInETH, "Assessment deposit is insufficient");
     if (msg.value > assessmentDepositInETH) {
       // Refund ETH excess back to the sender
       (
         bool refunded,
         /* bytes data */
-      ) = msg.sender.call{value: msg.value - assessmentDepositInETH}("");
+      ) = owner.call{value: msg.value - assessmentDepositInETH}("");
```
Comment (shark0der) [inline contracts/modules/assessment/IndividualClaims.sol:335]: Should use the `owner` instead of `msg.sender` when emitting the `ClaimSubmitted` as well.

File: contracts/modules/assessment/IndividualClaims.sol [hunk 22]
```diff
@@ -335,6 +352,7 @@ contract IndividualClaims is IIndividualClaims, MasterAwareV2 {
   /// When the tokens are transfered the assessment deposit is also sent back.
   ///
   /// @param claimId  Claim identifier
+
```
Comment (shark0der) [inline contracts/modules/assessment/IndividualClaims.sol:355]: Accidentally inserted new line?

**Review by danoctavian — APPROVED** [review]
LGTM
